File: python/ulsdk/test_context.py
# Copyright (c), CommunityLogiq Software

import uuid
from typing import Dict, List, Optional, Union
from websockets.sync.client import ClientConnection

from .api_key_context import ApiKeyContext
from .keys import Environment, Region
from .request_context import File, RequestContext


class TestContext(RequestContext):
    def __init__(self, context: ApiKeyContext):
        self._context = context
        self._response = b""

    def set_response(self, response: bytes):
        self._response = response

    def user_id(self) -> uuid.UUID:
        return self._context.user_id()

    def env(self) -> Environment:
        return self._context.env()

    def region(self) -> Region:
        return self._context.region()

    def get(
        self,
        path: str,
        params: Optional[Dict] = None,
        headers: Optional[Dict[str, str]] = None,
        **kwargs,
    ) -> bytes:
        echo_path = "/v1/echo/"
        self._context.get(echo_path, params, headers, **kwargs)
        return self._response

    def put(
        self,
        path: str,
        body: Union[bytes, str, None] = None,
        mimetype: str = "application/octet-stream",
        params: Optional[Dict] = None,
        headers: Optional[Dict[str, str]] = None,
        **kwargs,
    ) -> bytes:
        echo_path = "/v1/echo/"
        response = self._context.put(
            echo_path, self._response, mimetype, params, headers, **kwargs
        )
        if response != self._response:
            raise Exception("Test failure, expected response to match request")
        return self._response

    def post(
        self,
        path: str,
        body: Union[bytes, str, None] = None,
        mimetype: str = "application/octet-stream",
        params: Optional[Dict] = None,
        headers: Optional[Dict[str, str]] = None,
        **kwargs,
    ) -> bytes:
        echo_path = "/v1/echo/"
        response = self._context.post(
            echo_path, self._response, mimetype, params, headers, **kwargs
        )
        if response != self._response:
            raise Exception("Test failure, expected response to match request")
        return self._response

    def upload(
        self,
        path: str,
        files: List[File],
        params: Optional[Dict] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> bytes:
        echo_path = "/v1/echo/"
        self._context.upload(echo_path, files, params, headers)
        return self._response

    def delete(
        self,
        path: str,
        params: Optional[Dict] = None,
        headers: Optional[Dict[str, str]] = None,
        **kwargs,
    ) -> bytes:
        echo_path = "/v1/echo/"
        self._context.delete(echo_path, params, headers, **kwargs)
        return self._response

    def connect(
        self,
        path: str,
        params: Optional[Dict] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> ClientConnection:
        raise Exception("Websockets unsupported for tests")

File: go/test_context.go
// Copyright (c), CommunityLogiq Software

package ulsdk

import (
	"bytes"
	"fmt"
	"os"

	"github.com/google/uuid"
)

// TestContext wraps an ApiKeyContext and returns canned responses for testing.
// GET and DELETE requests are forwarded to /v1/echo/ to validate signing.
// POST and PUT requests send the canned response to /v1/echo/ and verify
// the echo matches, ensuring round-trip serialization correctness.
type TestContext struct {
	context  *ApiKeyContext
	response []byte
}

// NewTestContextFromEnv creates a TestContext using credentials from the
// CA_USER, CA_ACCESS_KEY, and CA_SECRET_KEY environment variables. It returns
// nil if any of the variables are unset, allowing callers to skip the test.
func NewTestContextFromEnv() *TestContext {
	user := os.Getenv("CA_USER")
	accessKey := os.Getenv("CA_ACCESS_KEY")
	secretKey := os.Getenv("CA_SECRET_KEY")
	if user == "" || accessKey == "" || secretKey == "" {
		return nil
	}
	userID, err := uuid.Parse(user)
	if err != nil {
		return nil
	}
	key := &Key{
		UserID:    userID,
		Region:    RegionCA,
		AccessKey: accessKey,
		SecretKey: secretKey,
	}
	keyCtx := NewApiKeyContext(key, EnvironmentStage)
	return NewTestContext(keyCtx)
}

// NewTestContext creates a new TestContext wrapping the given ApiKeyContext.
func NewTestContext(context *ApiKeyContext) *TestContext {
	return &TestContext{
		context:  context,
		response: []byte{},
	}
}

// SetResponse sets the canned response that will be returned by all requests.
func (c *TestContext) SetResponse(response []byte) {
	c.response = response
}

// Get makes an authenticated GET to /v1/echo/ and returns the canned response.
func (c *TestContext) Get(path string, params [][2]string, headers map[string]string) ([]byte, error) {
	_, err := c.context.Get("/v1/echo/", params, headers)
	if err != nil {
		return nil, err
	}
	return c.response, nil
}

// Post sends the canned response to /v1/echo/ and verifies the echo matches.
func (c *TestContext) Post(path string, body []byte, contentType string, params [][2]string, headers map[string]string) ([]byte, error) {
	r, err := c.context.Post("/v1/echo/", c.response, contentType, params, headers)
	if err != nil {
		return nil, err
	}
	if !bytes.Equal(r, c.response) {
		return nil, fmt.Errorf("test failure, expected response to match request")
	}
	return c.response, nil
}

// Put sends the canned response to /v1/echo/ and verifies the echo matches.
func (c *TestContext) Put(path string, body []byte, contentType string, params [][2]string, headers map[string]string) ([]byte, error) {
	r, err := c.context.Put("/v1/echo/", c.response, contentType, params, headers)
	if err != nil {
		return nil, err
	}
	if !bytes.Equal(r, c.response) {
		return nil, fmt.Errorf("test failure, expected response to match request")
	}
	return c.response, nil
}

// Delete makes an authenticated DELETE to /v1/echo/ and returns the canned response.
func (c *TestContext) Delete(path string, params [][2]string, headers map[string]string) ([]byte, error) {
	_, err := c.context.Delete("/v1/echo/", params, headers)
	if err != nil {
		return nil, err
	}
	return c.response, nil
}
